File: src/cc/Common/Filesystem.cc
/** -*- c++ -*-
 * Copyright (C) 2007-2012 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 3 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "Common/Compat.h"
#include "Common/Error.h"
#include "Common/Serialization.h"

#include "Filesystem.h"

using namespace Hypertable;
using namespace Serialization;

int Filesystem::decode_response_open(EventPtr &event_ptr) {
  const uint8_t *decode_ptr = event_ptr->payload;
  size_t decode_remain = event_ptr->payload_len;
  int error = decode_i32(&decode_ptr, &decode_remain);

  if (error != Error::OK)
    HT_THROW(error, "");

  return decode_i32(&decode_ptr, &decode_remain);
}


int Filesystem::decode_response_create(EventPtr &event_ptr) {
  return decode_response_open(event_ptr);
}


size_t
Filesystem::decode_response_read(EventPtr &event_ptr, void *dst, size_t len) {
  const uint8_t *decode_ptr = event_ptr->payload;
  size_t decode_remain = event_ptr->payload_len;

  int error = decode_i32(&decode_ptr, &decode_remain);

  if (error != Error::OK)
    HT_THROW(error, "");

  // uint64_t offset =
  decode_i64(&decode_ptr, &decode_remain);
  uint32_t nread = decode_i32(&decode_ptr, &decode_remain);

  if (nread == (uint32_t)-1)
    return 0;

  if (decode_remain < nread)
    HT_THROWF(Error::RESPONSE_TRUNCATED, "%lu < %lu",
              (Lu)decode_remain, (Lu)nread);

  // PERF: We could just send back a pointer to msg here
  memcpy(dst, decode_ptr, nread);

  return nread;
}


/**
 */
size_t
Filesystem::decode_response_read_header(EventPtr &event_ptr, uint64_t *offsetp,
                                        uint8_t **dstp) {
  const uint8_t *decode_ptr = event_ptr->payload;
  size_t decode_remain = event_ptr->payload_len;

  int error = decode_i32(&decode_ptr, &decode_remain);

  if (error != Error::OK)
    HT_THROW(error, "");

  *offsetp = decode_i64(&decode_ptr, &decode_remain);
  uint32_t amount = decode_i32(&decode_ptr, &decode_remain);

  if (dstp)
    *dstp = (uint8_t *)decode_ptr;

  return amount;
}


/**
 */
size_t
Filesystem::decode_response_pread(EventPtr &event_ptr, void *dst, size_t len) {
  return decode_response_read(event_ptr, dst, len);
}


/**
 */
size_t
Filesystem::decode_response_append(EventPtr &event_ptr, uint64_t *offsetp) {
  const uint8_t *decode_ptr = event_ptr->payload;
  size_t decode_remain = event_ptr->payload_len;

  int error = decode_i32(&decode_ptr, &decode_remain);

  if (error != Error::OK)
    HT_THROW(error, "");

  *offsetp = decode_i64(&decode_ptr, &decode_remain);

  return decode_i32(&decode_ptr, &decode_remain);
}


/**
 */
int64_t
Filesystem::decode_response_length(EventPtr &event_ptr) {
  const uint8_t *decode_ptr = event_ptr->payload;
  size_t decode_remain = event_ptr->payload_len;

  int error = decode_i32(&decode_ptr, &decode_remain);

  if (error != Error::OK)
    HT_THROW(error, "");

  return decode_i64(&decode_ptr, &decode_remain);
}


/**
 */
void
Filesystem::decode_response_readdir(EventPtr &event_ptr,
                                    std::vector<String> &listing) {
  const uint8_t *decode_ptr = event_ptr->payload;
  size_t decode_remain = event_ptr->payload_len;

  int error = decode_i32(&decode_ptr, &decode_remain);

  if (error != Error::OK)
    HT_THROW(error, "");

  listing.clear();

  uint32_t len = decode_i32(&decode_ptr, &decode_remain);
  uint16_t slen;

  for (uint32_t i=0; i<len; i++) {
    const char *str = decode_str16(&decode_ptr, &decode_remain, &slen);
    listing.push_back(String(str, slen));
  }
}


/**
 */
bool
Filesystem::decode_response_exists(EventPtr &event_ptr) {
  const uint8_t *decode_ptr = event_ptr->payload;
  size_t decode_remain = event_ptr->payload_len;

  int error = decode_i32(&decode_ptr, &decode_remain);

  if (error != Error::OK)
    HT_THROW(error, "");

  return decode_bool(&decode_ptr, &decode_remain);
}



/**
 */
int
Filesystem::decode_response(EventPtr &event_ptr) {
  const uint8_t *decode_ptr = event_ptr->payload;
  size_t decode_remain = event_ptr->payload_len;

  return decode_i32(&decode_ptr, &decode_remain);
}

static String &
remove_trailing_duplicates(String &s, char separator)
{
  while (s.size() > 1) {
    if (s[s.size()-1]==separator)
      s.resize(s.size()-1);
    else
      break;
  }

  return s;
}

/** 
 */
String Filesystem::dirname(String name, char separator)
{
  if (name.size() == 1 && name[0] == separator)
    return name;

  // Remove trailing separators
  std::string::size_type cur = name.length();
  while (cur > 0 && name[cur - 1] == separator)
    --cur;
  name.resize(cur);

  // Find last separator
  std::string::size_type pos = name.rfind(separator);

  if (pos == std::string::npos)
    name = "."; // No separators
  else if (pos == 0 && name.length() == 1 && name[0] == separator)
    name = separator; // Only separators
  else if (pos == 0 && name.length() > 1 && name[0] == separator)
    name = "/";
  else
    name.resize(pos);

  // Remove any duplicate adjacent path separators
  return remove_trailing_duplicates(name, separator);
}

/** 
 */
String Filesystem::basename(String name, char separator)
{
  if (name.size() == 1 && name[0] == separator)
    return name;

  // Remove trailing separators
  std::string::size_type cur = name.length();
  while (cur > 0 && name[cur - 1] == separator)
    --cur;
  name.resize(cur);

  // Find last separator
  std::string::size_type pos = name.rfind(separator);

  std::string ret;
  if (pos == std::string::npos)
    ; // No separators
  else if (pos == 0 && name.length() == 1 && name[0] == separator)
    name = separator; // Only separators
  else
    name = name.substr(pos + 1); // Basename only

  // Remove any duplicate adjacent path separators
  return remove_trailing_duplicates(name, separator);
}

File: java/hypertable-common/src/main/java/org/hypertable/Common/Filesystem.java
/*
 * Copyright (C) 2007-2015 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

package org.hypertable.Common;

import java.io.UnsupportedEncodingException;
import java.net.ProtocolException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.zip.DataFormatException;

public class Filesystem {

  /** Enumeration for flags */
  public enum Flags {
    NONE(0), FLUSH(1), SYNC(2);
    private final int value;
    private Flags(int value) {
      this.value = value;
    }
    public int getValue() {
      return value;
    }
    private static Flags[] allValues = values();
    public static Flags fromOrdinal(int n) {return allValues[n];}
  };

  /**
   * A directory entry for readdir.
   */
  public static class Dirent {
    static final byte DIRENT_VERSION = 1;
    public int EncodedLength() throws UnsupportedEncodingException {
      int length = 13 + Serialization.EncodedLengthVStr(name);
      return 1 + Serialization.EncodedLengthVInt32(length) + length;
    }
    public void Encode(ByteBuffer buf) throws UnsupportedEncodingException {
      buf.order(ByteOrder.LITTLE_ENDIAN);
      buf.put(DIRENT_VERSION);
      int encoding_length = 13 + Serialization.EncodedLengthVStr(name);
      Serialization.EncodeVInt32(buf, encoding_length);
      Serialization.EncodeVStr(buf, name);
      buf.putLong(length);
      buf.putInt(last_modification_time);
      buf.put(is_dir ? (byte)1 : (byte)0);
    }
    public void Decode(ByteBuffer buf)
      throws ProtocolException, UnsupportedEncodingException, DataFormatException {
      buf.order(ByteOrder.LITTLE_ENDIAN);
      if (buf.remaining() < 2)
        throw new ProtocolException("Truncated message");
      int version = (int)buf.get();
      if (version != DIRENT_VERSION)
        throw new ProtocolException("Dirent parameters version mismatch, expected " +
                                    DIRENT_VERSION + ", got " + version);
      int encoding_length = Serialization.DecodeVInt32(buf);
      int start_position = buf.position();
      name = Serialization.DecodeVStr(buf);
      length = buf.getLong();
      last_modification_time = buf.getInt();
      is_dir = buf.get() == (byte)0 ? false : true;
      if ((buf.position() - start_position) < encoding_length)
        buf.position(start_position + encoding_length);
    }
    public String name;
    public long length = 0;
    public int last_modification_time = 0;
    public boolean is_dir = false;
  }

}
